pkg/reconciler/buildrun/resources/taskrun.go

```diff
@@ -319,6 +347,36 @@ func GenerateTaskRun(
 	}
 
 	expectedTaskRun.Spec.Params = params
+
+	// Ensure a proper override of params between Build and BuildRun
+	// A BuildRun can override a param as long as it was defined in the Build
+	buildUserParams := OverrideParams(build.Spec.Params, buildRun.Spec.Params)
+
+	// list of params that collide with reserved system strategy parameters
+	undesiredParams := []string{}
+
+	// Append params to the TaskRun spec definition
+	for _, p := range buildUserParams {
+
+		if m := systemReservedParametersRegexp.MatchString(p.Name); m {
+			undesiredParams = append(undesiredParams, p.Name)
+		}
+
+		buildParam := v1beta1.Param{
+			Name: p.Name,
+			Value: v1beta1.ArrayOrString{
+				Type:      v1beta1.ParamTypeString,
+				StringVal: p.Value,
+			},
+		}
+		expectedTaskRun.Spec.Params = append(expectedTaskRun.Spec.Params, buildParam)
+	}
+	// if system parameters names are being use, fail the taskRun creation and update the condition message
+	// with a custom error
+	if len(undesiredParams) > 0 {
+		return nil, fmt.Errorf("restricted parameters in use: %s", strings.Join(undesiredParams, ","))
+	}
+
 	return expectedTaskRun, nil
```

Comment:
As we allow build strategy parameters without a default (see earlier comment), we would need to check for those that the Build or BuildRun provides a value for them, otherwise I guess Tekton will not allow the TaskRun to be created?

Reply:
name and value are required fields in the `params` object of both Build/BuildRun, so k8s will complain of that. No need to check that at this level.

Example:

```sh
error: error validating "samples/buildrun/buildrun_buildkit_cr.yaml": error validating data: ValidationError(BuildRun.spec.params[0]): missing required field "value" in io.shipwright.v1alpha1.BuildRun.spec.params; if you choose to ignore these errors, turn validation off with --validate=false
´´`

Reply:
Hi @qu1queee, I meant something else. Quickly put these changes into this commit: https://github.com/SaschaSchwarze0/build/commit/29eccc4211390bc958a805f3426929c29d12c1cc:

* Change `default` to `*string` and remove `omitempty`: this allows a build strategy author to omit the default to indicate that the Build or BuildRun MUST specify that param. It also allows the author to define `""` as a default.
* Change the TaskSpec generation to omit the whole `Default` inside the `ParamSpec` when the `default` in the build strategy is `nil`.
* Added a check at the TaskRun generation: if there is any build strategy parameter without default that is not specified in the Build or BuildRun, then the error `"value for parameters missing: %s", strings.Join(missingParams, ",")` is set for the BuildRun
* Added integration tests for strategy with parameter without `default` and with `default` set to `""`

Reply:
done, thanks, this was helpful